### PR TITLE
Add a setting allowing to throttle ANALYZE

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Relates to https://github.com/crate/crate/issues/12470

### CODE REVIEW

**Comment by BaurzhanSakhariev** [comment]
Hi @seut, could you please provide an early feedback?

No docs/tests on this stage.

As discussed, a new field in the request is needed - to differentiate whether it's manual or auto analyze and only then extend grammar and pass NULL from manual call. 

However, it seems automatic run doesn't use directly some internal code as I thought before, it calls `ANALYZE` statement
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/statistics/TableStatsService.java#L111

so we need extend grammar already - I will do it in the follow up commits once this general approach is fine.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 1]
```diff
@@ -89,20 +93,59 @@ public final class ReservoirSampler {
...
+        boolean isAnalyzing = analyzing; // One volatile read.
+        if (isAnalyzing == false) {
+            readLimit = newReadLimit;
+            rateLimiter.setMBPerSec(readLimit.getMbFrac());
+        } else {
+            updatedReadLimit = newReadLimit; // One volatile write. Let running ANALYZE operation finish with the previous value.
+        }
+    }
+
+    public void maybePause(long bytes) {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:135]: this is derived from https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/index/snapshots/blobstore/RateLimitingInputStream.java#L49

can actually refactor/reuse if this approach is fine in general

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 2]
```diff
@@ -89,20 +93,59 @@ public final class ReservoirSampler {
...
+        bytesSinceLastRateLimit += bytes;
+        if (bytesSinceLastRateLimit >= rateLimiter.getMinPauseCheckBytes()) {
+            try {
```
Comment (seut) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:138]: shouldn't the bytes have to be subtracted in this case gain? wouldn't calling it `bytesSinceLastPause` express it in a better way?

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 3]
```diff
@@ -89,20 +93,59 @@ public final class ReservoirSampler {
...
+        this.updatedReadLimit = readLimit;
+        this.analyzing = false;
+
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(
+            STATS_SERVICE_THROTTLING_SETTING, this::updateReadLimit);
+    }
+
+    private void updateReadLimit(ByteSizeValue newReadLimit) {
+        boolean isAnalyzing = analyzing; // One volatile read.
+        if (isAnalyzing == false) {
```
Comment (seut) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:127]: why can't we adjust the RateLimiter directly? Why do we have to wait until a running analyze finishes?

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 4]
```diff
@@ -89,20 +93,59 @@ public final class ReservoirSampler {
     private CircuitBreakerService circuitBreakerService;
     private final IndicesService indicesService;
 
+    private volatile ByteSizeValue readLimit;
+    private volatile ByteSizeValue updatedReadLimit;
+    private volatile boolean analyzing;
+
+    private final RateLimiter rateLimiter;
+    private long bytesSinceLastRateLimit = 0;
```
Comment (seut) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:101]: As this variable could be adjusted by different threads, we must ensure thread-safety here.
```suggestion
    private AtomicLong bytesSinceLastRateLimit = new AtomicLong();
```

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 5]
```diff
@@ -89,20 +93,59 @@ public final class ReservoirSampler {
...
+    }
+
+    private void updateReadLimit(ByteSizeValue newReadLimit) {
+        boolean isAnalyzing = analyzing; // One volatile read.
+        if (isAnalyzing == false) {
+            readLimit = newReadLimit;
+            rateLimiter.setMBPerSec(readLimit.getMbFrac());
+        } else {
+            updatedReadLimit = newReadLimit; // One volatile write. Let running ANALYZE operation finish with the previous value.
+        }
```
Comment (seut) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:132]: I suggest to set the rateLimiter to NULL to disable it to avoid useless calls.

**Review by seut — COMMENTED** [review]
Look good to me in general, added some suggestions, maybe to detailed for a draft, sorry :)

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 6]
```diff
@@ -89,20 +93,59 @@ public final class ReservoirSampler {
...
+            STATS_SERVICE_THROTTLING_SETTING, this::updateReadLimit);
+    }
+
+    private void updateReadLimit(ByteSizeValue newReadLimit) {
+        boolean isAnalyzing = analyzing; // One volatile read.
+        if (isAnalyzing == false) {
+            readLimit = newReadLimit;
+            rateLimiter.setMBPerSec(readLimit.getMbFrac());
+        } else {
+            updatedReadLimit = newReadLimit; // One volatile write. Let running ANALYZE operation finish with the previous value.
```
[the remark this answers is not in the document]
Comment (seut) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java:131]: Yes I see no reason to not update a running analyze, I think this is even wanted like adjust it as the current analyze is slow and there is currently enough headroom to run it faster.

**Comment by seut** [comment]
> However, it seems automatic run doesn't use directly some internal code as I thought before, it calls ANALYZE statement
> https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/statistics/TableStatsService.java#L111
>
>so we need extend grammar already - I will do it in the follow up commits once this general approach is fine.

Yes right, fine for me to do this afterwards (so we can disable throttling for explicit ANALYZE calls by default).

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 7]
```diff
@@ -190,7 +233,9 @@ private Samples getSamples(List<Reference> columns,
                     Engine.Searcher searcher = indexShard.acquireSearcher("update-table-statistics");
                     searchersToRelease.add(searcher);
                     totalNumDocs += searcher.getIndexReader().numDocs();
-                    totalSizeInBytes += indexShard.storeStats().getSizeInBytes();
+                    long bytesRead = indexShard.storeStats().getSizeInBytes();
+                    totalSizeInBytes += bytesRead;
+                    maybePause(bytesRead);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:238]: I think most of the time will be spent in the `ReservoirCollector` and further below where we load the values from disk: 

https://github.com/crate/crate/blob/a3eb7f078f1e7981b7489e03ecf93d027bb575b0/server/src/main/java/io/crate/statistics/ReservoirSampler.java#L214-L215

Only throttling up here is probably not having a large impact.

File: server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java [hunk 8]
```diff
@@ -211,11 +199,9 @@ static Stats createTableStats(Samples samples, List<Reference> primitiveColumns)
 
     private CompletableFuture<Samples> fetchSamples(RelationName relationName, List<Reference> columns) {
         FutureActionListener<FetchSampleResponse, Samples> listener = new FutureActionListener<>(FetchSampleResponse::samples);
-        List<DiscoveryNode> nodesOn41OrAfter = StreamSupport.stream(clusterService.state().nodes().spliterator(), false)
-            .filter(x -> x.getVersion().onOrAfter(Version.V_4_1_0))
-            .collect(Collectors.toList());
+        DiscoveryNodes discoveryNodes = clusterService.state().nodes();
         MultiActionListener<FetchSampleResponse, ?, FetchSampleResponse> multiListener = new MultiActionListener<>(
-            nodesOn41OrAfter.size(),
+            discoveryNodes.getSize(),
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java:204]: Could we split these changes out into a dedicated PR?

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mfussenegger) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java:131]: I think it would actually be acceptable to _not_ update the throttling value of a running analyze, but no strong opinion.

But it would have the advantage that we could instantiate the RateLimiter on a new call with the current value from the cluster setting. No need for an extra setting-update-consumer or keeping track of multiple volatile variables.

**Comment by mfussenegger** [comment]
> Yes right, fine for me to do this afterwards (so we can disable throttling for explicit ANALYZE calls by default).

I'd keep things simple and not distinguish between manual ANALYZE and job stats calls and throttle both.
As a user you probably also don't want to impact query performance too much if you invoke it manually.

File: server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java (hunk 8 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java:204]: https://github.com/crate/crate/pull/13052

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 4 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java]: I think it actually can be even simpler: bytes counter is relevant per-analyse, idk why I declared it as a class field in the first place. I use it now per sampling in https://github.com/crate/crate/pull/13046/commits/ef4e45b308401b4851e502b5a1d80b3b1d6d24d1.

Updated counter name as well.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 1 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java]: wouldn't touch `RateLimitingInoutStream` as I changed `maybePause` in https://github.com/crate/crate/commit/ef4e45b308401b4851e502b5a1d80b3b1d6d24d1 to accept current state in order to avoid counter synchronization.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java]: I let to update on the fly in https://github.com/crate/crate/pull/13046/commits/ebd3d49754ebdf208efc5c59a8fb801e692d5c31

motivation is to let users adjust the new setting optimally for their environment/setup. Would also mention in docs smth like 

 _"setting can be adjusted optimally for the concrete setup by running ANALYZE and playing around with the setting and monitoring cluster state at the same time"_
 
 currently there is no volatile variables other than `mbPerSec` which is used internally in `SimpleRateLimiter`

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 3 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java]: replied in https://github.com/crate/crate/pull/13046#discussion_r976408756

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 5 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java]: covered disabled case in https://github.com/crate/crate/pull/13046/commits/a15139bc2494b2c000c8c4ecf20c3eb624988367
I didn't set limiter to null since it might be updated in `updateReadLimit` and can cause NPE. 
Currently maubepause is not called if setting is disabled - later on when we extend grammar, getSamles will get additional flag and it will be also aded to the check.

**Comment by BaurzhanSakhariev** [comment]
> As a user you probably also don't want to impact query performance too much if you invoke it manually.

if we extend grammar, user can do `throttle = on` and use the setting. We would have more flexibility with it - but no strong opinion, can always throttle both auto/manual

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 7 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java]: To elaborate on this - if we do the check based on the size of the indexShard we'll have _one_ large pause per shard and still cause large IO spikes. What we want is to smooth out the IO access.

For that we'd need to account bytes read *per row* and pause during the inner loop.

**Comment by mfussenegger** [comment]
> if we extend grammar, user can do throttle = on and use the setting. We would have more flexibility with it - but no strong opinion, can always throttle both auto/manual

It's easy to extend this later if needed. If it isn't needed it causes unecessary complexity.
As a workaround a user can increase the rate-limit before invoking ANALYZE and reset it afterwards.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 7 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java]: thx for the details, done in https://github.com/crate/crate/pull/13046/commits/a25da8cbe3f3c6dfb74fc12f2a32c94149426c3e

**Comment by BaurzhanSakhariev** [comment]
Hi @seut, @mfussenegger could you please have another look? 

TODO: restrict triggering `ANALYZE` if scheduled one is in progress.

Current implementation allows to update setting when `ANALYZE` is in progress - it can be used to adjust the setting by triggering `ANALYZE` and playing around with the limit.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 9]
```diff
@@ -213,8 +243,16 @@ private Samples getSamples(List<Reference> columns,
...
             try {
-                rowAccounting.accountForAndMaybeBreak(row);
+                long bytesRead = rowAccounting.accountRowBytes(row);
+                ramAccounting.addBytes(bytesRead);
+                totalSizeInBytes += bytesRead;
+                long minPauseCheckBytes = rateLimiter.getMinPauseCheckBytes(); // volatile read, gets updated together with rateLimiter.mbPerSec
+                if (minPauseCheckBytes > 0) {
+                    // Throttling is enabled
+                    bytesSinceLastPause = maybePause(bytesRead, bytesSinceLastPause);
+                }
```
Comment (seut) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:255]: minor, but I'd prefer all maybe pause logic to be at one place, so I'd suggest to move that inside `maybePause`.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 10]
```diff
@@ -213,8 +243,16 @@ private Samples getSamples(List<Reference> columns,
...
             DocIdToRow docIdToRow = docIdToRowsFunctionPerReader.get(readerId);
             Object[] row = docIdToRow.apply(FetchId.decodeDocId(fetchId));
+
             try {
-                rowAccounting.accountForAndMaybeBreak(row);
+                long bytesRead = rowAccounting.accountRowBytes(row);
+                ramAccounting.addBytes(bytesRead);
+                totalSizeInBytes += bytesRead;
+                long minPauseCheckBytes = rateLimiter.getMinPauseCheckBytes(); // volatile read, gets updated together with rateLimiter.mbPerSec
+                if (minPauseCheckBytes > 0) {
```
Comment (seut) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:252]:
```suggestion
                if (rate.getMBPerSec() > 0) {
```

**Review by seut — COMMENTED** [review]
pattern looks good to me, added few suggestions.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 11]
```diff
@@ -213,8 +243,16 @@ private Samples getSamples(List<Reference> columns,
             int readerId = FetchId.decodeReaderId(fetchId);
             DocIdToRow docIdToRow = docIdToRowsFunctionPerReader.get(readerId);
             Object[] row = docIdToRow.apply(FetchId.decodeDocId(fetchId));
+
             try {
-                rowAccounting.accountForAndMaybeBreak(row);
```
Comment (matriv) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:217]: Is it correct that we have removed this call? Seems that now we don't have a circuitbreaker exception if accounting exceeds memory limits.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 9 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java]: moved in https://github.com/crate/crate/pull/13046/commits/94e60d3014fc2ba2bf6cb5a4c9a5021e55465fae

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 11 above)
Comment (matriv) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java:217]: sorry sorry, I somehow missed it. Thx!

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 7 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java]: commit ^ had a bug - I acidentally added row size to `totalSizeInBytes`. In the master branch we add only `indexShard` stats size.

 Pushed a fixup https://github.com/crate/crate/pull/13046/commits/1daf8d8cd6320e299e2f9afacbb8ddd938fadb2a but it's a bit worrying that all tests passed even with buggy version. 
 
Looks like `totalSizeInBytes` computed in `ReservoirSampler` is not used in `pg_stats` and only used in some optimizers - so it's probably fine that there is no visible effect of that bug.
 
 `Samples.totalSizeInBytes` is used in `Stats.sizeInBytes` which is indirectly used **only** in `Stats.averageSizePerRowInBytes` (`Stats.sizeInBytes` is unused directly somewhere else) ==> this is used in couple of optimisers. I cleaned up unused code in https://github.com/crate/crate/pull/13046/commits/59cab831ca4dc56f6eb23e74c0ebb80ff3f4dba1

**Comment by BaurzhanSakhariev** [comment]
> TODO: restrict triggering `ANALYZE` if scheduled one is already in progress.

this is not related to throttling and will be done in a follow up PR

**Comment by BaurzhanSakhariev** [comment]
retest this please

File: docs/sql/statements/analyze.rst [hunk 12]
```diff
@@ -28,3 +28,9 @@ plans.
 
 The statistics are also periodically updated. How often can be configured with
 the :ref:`stats.service.interval <stats.service.interval>` setting.
+
+Statistics collection IO can be throttled with the
```
Comment (matriv) [inline docs/sql/statements/analyze.rst:32]:
```suggestion
I/O throughput during collection of statistics can be throttled with the
```

File: docs/sql/statements/analyze.rst [hunk 13]
```diff
@@ -28,3 +28,9 @@ plans.
...
+:ref:`stats.service.max_bytes_per_sec <stats.service.max_bytes_per_sec>`
+setting. Setting can be applied when analysis is in progress and thus it's
+possible to adjust optimal value for a concrete setup by running ``ANALYZE``
```
[the remark this answers is not in the document]
Comment (matriv) [reply docs/sql/statements/analyze.rst:35]: Myself I like this sentence to be included but I'd rephrase it a bit:
```
Changes to this setting can be made and take effect while an analysis is in progress, this it's possible...
```

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 14]
```diff
@@ -89,17 +94,59 @@ public final class ReservoirSampler {
...
+        this(clusterService,
+             nodeCtx,
+             schemas,
+             circuitBreakerService,
+             indicesService,
+             new RateLimiter.SimpleRateLimiter(STATS_SERVICE_THROTTLING_SETTING.get(settings).getMbFrac())
+        );
+    }
+
+    public ReservoirSampler(ClusterService clusterService,
```
Comment (matriv) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:115]: Can this ctor can be package private and `@VisibleForTesting`?

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 15]
```diff
@@ -226,9 +290,12 @@ private Samples getSamples(List<Reference> columns,
             }
             records.add(new RowN(row));
         }
-        return new Samples(records, streamers, totalNumDocs, totalSizeInBytes);
+
+        return records;
     }
 
+
```
Comment (matriv) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:297]: two many empty lines :) :) ?

**Review by matriv — COMMENTED** [review]
Looks good! Left some comments/suggestions around docs,
If we use IO instead of I/O elsewhere, no need to change it, just seems better in my eyes.

File: docs/appendices/release-notes/unreleased.rst [hunk 16]
```diff
@@ -109,6 +109,9 @@ Changes
 - Updated to Admin UI 1.23.0, which improves scrolling behavior on wide result
   sets.
 
+- Added a new setting ``stats.service.max_bytes_per_sec`` which allows to
+  throttle IO intensity of the ``ANALYZE`` statement.
```
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply docs/appendices/release-notes/unreleased.rst:113]: throughput sounds good to me

**Review by matriv — APPROVED** [review]
LGTM, please wait for at least one of the 2 other reviewers as well.

File: docs/config/cluster.rst [hunk 17]
```diff
@@ -179,6 +179,18 @@ Collecting stats
 
     Using a very small value can cause a high load on the cluster.
 
+.. _stats.service.max_bytes_per_sec:
+
+**stats.service.max_bytes_per_sec**
+  | *Default:*    ``0``
+  | *Runtime:*   ``yes``
```
Comment (mfussenegger) [inline docs/config/cluster.rst:186]: Given that part of the motivation for this change is that users experience unexpected performance hiccups due to the analyze I think we should enable the throttling by default.

Not sure what a good value would be, probably worth running some tests on a lower specced cluster and aim for a value that errors on the safe-but-maybe-slow side of things?

Comment (BaurzhanSakhariev) [reply docs/config/cluster.rst:186]: Enabling by default sounds good to me.

On choosing default - maybe we can align with existing` *_bytes_per_sec` settings - would make it `40mb` which looks like a reasonable lower bound (just randomly looked up one of the I/O throttling  `_bytes_per_sec` settings and found a [table](https://www.elastic.co/guide/en/elasticsearch/reference/master/recovery.html#recovery-settings)).

Comment (mfussenegger) [reply docs/config/cluster.rst:186]: 40 sounds good, but some manual testing to see the impact would be nice anyways.

Comment (BaurzhanSakhariev) [reply docs/config/cluster.rst]: Hi @mfussenegger, I changed default to 40mb and checked on dev cluster:
https://tools.cr8.net/grafana/d/hkhCfqdmk/dev-cluster?var-node=All&from=now-3h&to=now 


I run ANALYZE 4 times with the following values:` 0, 40mb, 40mb, 0` - latency got higher with enabled throttling as expected.

Screenshot 2022-09-28 at 14 08 41

Comment (BaurzhanSakhariev) [reply docs/config/cluster.rst]: Chart for `node_disk_read_bytes_total`, run with 0, 20, 60 - the higher value I set for throttling, the lower IOPS gets.

Screenshot 2022-09-28 at 16 30 37

File: docs/appendices/release-notes/unreleased.rst [hunk 18]
```diff
@@ -112,6 +112,9 @@ Changes
 - Updated to Admin UI 1.23.0, which improves scrolling behavior on wide result
   sets.
 
+- Added a new setting ``stats.service.max_bytes_per_sec`` which allows to
+  throttle I/O throughput of the ``ANALYZE`` statement.
```
Comment (seut) [inline docs/appendices/release-notes/unreleased.rst:116]:
```suggestion
- Added I/O throughput throttling of the :ref:`analyze` statement as
  well as of the periodic statistic collection controlled by the 
  :ref:`stats.service.interval` setting to lower the impact on the
  cluster load. This throttling can be controlled by a new setting
  :ref:`stats.service.max_bytes_per_sec` and is set 40MB/s by
  default.
```

File: docs/sql/statements/analyze.rst [hunk 19]
```diff
@@ -28,3 +28,9 @@ plans.
 
 The statistics are also periodically updated. How often can be configured with
 the :ref:`stats.service.interval <stats.service.interval>` setting.
+
+I/O throughput during collection of statistics can be throttled with the
+:ref:`stats.service.max_bytes_per_sec <stats.service.max_bytes_per_sec>`
```
Comment (seut) [inline docs/sql/statements/analyze.rst:33]:
```suggestion
:ref:`stats.service.max_bytes_per_sec`
```

File: server/src/test/java/io/crate/statistics/ReservoirSamplerTest.java [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+            createNodeContext(),
+            mock(Schemas.class),
+            new NoneCircuitBreakerService(),
+            mock(IndicesService.class),
+            rateLimiter
+        );
+    }
+
+    @Test
+    public void default_throttling_disabled_rate_limiter_pause_not_called() throws IOException {
```
Comment (seut) [inline server/src/test/java/io/crate/statistics/ReservoirSamplerTest.java:70]:
```suggestion
    public void test_rate_limiter_pause_not_called_if_throttling_disabled() throws IOException {
```

File: server/src/test/java/io/crate/statistics/ReservoirSamplerTest.java [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+
+        RowCellsAccountingWithEstimators rowAccounting = mock(RowCellsAccountingWithEstimators.class);
+
+        // FetchId.decodeReaderId(123) returns 0, second List should have an element with index 0.
+        sampler.createRecords(List.of(fetchId), List.of(docIdToRow), RamAccounting.NO_ACCOUNTING, rowAccounting, 1);
+        verify(rateLimiter, never()).pause(anyLong());
+    }
+
+    @Test
+    public void throttling_enabled_rate_limiter_called() throws Exception {
```
Comment (seut) [inline server/src/test/java/io/crate/statistics/ReservoirSamplerTest.java:83]:
```suggestion
    public void test_rate_limiter_pause_called_if_throttling_enabled() throws Exception {
```

**Review by seut — COMMENTED** [review]
Added some doc & test suggestions

**Comment by BaurzhanSakhariev** [comment]
> Added some doc & test suggestions

I pushed a fixup https://github.com/crate/crate/pull/13046/commits/6edc3517ea035ab63697a33e68e3f1a29ca189dd - I think we cannot refer to not yet merged doc?

**Comment by seut** [comment]
@BaurzhanSakhariev 

>  I think we cannot refer to not yet merged doc?

Of course you can, you can (and also please should) validate your documentation/CHANGES entry changes by building the documentation locally via `./gradlew buildDocs` & open it in your browser.

**Comment by BaurzhanSakhariev** [comment]
> > I think we cannot refer to not yet merged doc?
> 
> Of course you can, you can (and also please do) validate your documentation/CHANGES entry changes by building the documentation locally via `./gradlew buildDocs` & open it in your browser.

thx, pushed a fixup. I had to use full ref with label name, when I use

:ref:`stats.service.max_bytes_per_sec`,  `buildDocs` fails with `Failed to create a cross reference. A title or caption not found: stats.service.max_bytes_per_sec`. 

I thought it's because of dots but  `:ref:`stats.service.interval`` above works and it contains dots.

**Review by seut — APPROVED** [review]
👍

**Comment by BaurzhanSakhariev** [comment]
retest this please

unrelated failure https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/4009/testReport/junit/io.crate.integrationtests/GroupByAggregateTest/test_select_distinct_with_limit_and_offset_applies_limit_and_offset_on_distinct_resultset/ similar to https://wacklig.pipifein.dev/github/crate/crate/c6173120-52cb-4ff6-876f-008c08ab3064/0c331b92-25ae-4493-a10c-97d453040390

looks like a new flaky but has to be addressed separately
